File: apps/etterlatte-saksbehandling-ui/client/src/shared/types/Etteroppgjoer.ts
import { ISak } from '~shared/types/sak'
import { IAvkortetYtelse, IAvkortingGrunnlag } from '~shared/types/IAvkorting'
import { GrunnlagKilde } from '~shared/types/grunnlag'

export interface Etteroppgjoer {
  behandling: EtteroppgjoerBehandling
  opplysninger: EtteroppgjoerOpplysninger
  faktiskInntekt?: FaktiskInntekt
  avkortingFaktiskInntekt: Avkorting | undefined
  beregnetEtteroppgjoerResultat: BeregnetEtteroppgjoerResultatDto | undefined
}

export interface EtteroppgjoerBehandling {
  id: string
  status: EtteroppgjoerBehandlingStatus
  relatertForbehandlingId: string // burde bruke denne for å hente faktiskInntekt etteroppgjør revurdering
  sak: ISak
  aar: number
  innvilgetPeriode: {
    fom: string
    tom: string
  }
  opprettet: string // Mottatt?
  brevId?: number
}

export enum EtteroppgjoerBehandlingStatus {
  OPPRETTET = 'OPPRETTET',
  BEREGNET = 'BEREGNET',
  FERDIGSTILT = 'FERDIGSTILT',
}

export const teksterEtteroppgjoerBehandlingStatus: Record<EtteroppgjoerBehandlingStatus, string> = {
  OPPRETTET: 'Opprettet',
  BEREGNET: 'Beregnet',
  FERDIGSTILT: 'Ferdigstilt',
}

export interface EtteroppgjoerOpplysninger {
  skatt: PensjonsgivendeInntektFraSkatteetaten
  ainntekt: AInntekt
  tidligereAvkorting: Avkorting
}

export interface FaktiskInntekt {
  loennsinntekt: number
  afp: number
  naeringsinntekt: number
  utland: number
  spesifikasjon: string
}

export interface PensjonsgivendeInntektFraSkatteetaten {
  inntekter: PensjonsgivendeInntekt[]
}

export interface PensjonsgivendeInntekt {
  skatteordning: string
  loensinntekt: number
  naeringsinntekt: number
  annet: number
}

export interface AInntekt {
  inntektsmaaneder: AInntektMaaned[]
}

export interface AInntektMaaned {
  maaned: string
  summertBeloep: number
}

export interface Avkorting {
  avkortingGrunnlag: IAvkortingGrunnlag[]
  avkortetYtelse: IAvkortetYtelse[]
}

export enum EtteroppgjoerResultatType {
  TILBAKEKREVING = 'TILBAKEKREVING',
  ETTERBETALING = 'ETTERBETALING',
  IKKE_ETTEROPPGJOER = 'IKKE_ETTEROPPGJOER',
}

export interface BeregnetEtteroppgjoerResultatDto {
  id: string
  aar: number
  forbehandlingId: string
  sisteIverksatteBehandlingId: string
  utbetaltStoenad: number
  nyBruttoStoenad: number
  differanse: number
  grense: EtteroppgjoerGrenseDto
  resultatType: EtteroppgjoerResultatType
  tidspunkt: string
  kilde: GrunnlagKilde
  avkortingForbehandlingId: string
  avkortingSisteIverksatteId: string
}

interface EtteroppgjoerGrenseDto {
  tilbakekreving: number
  etterbetaling: number
  rettsgebyr: number
  rettsgebyrGyldigFra: string
}

File: libs/etterlatte-beregning-model/src/main/kotlin/Etteroppgjoer.kt
package no.nav.etterlatte.libs.common.beregning

import no.nav.etterlatte.libs.common.grunnlag.Grunnlagsopplysning
import no.nav.etterlatte.libs.common.tidspunkt.Tidspunkt
import java.time.LocalDate
import java.util.UUID

data class BeregnetEtteroppgjoerResultatDto(
    val id: UUID,
    val aar: Int,
    val forbehandlingId: UUID,
    val sisteIverksatteBehandlingId: UUID,
    val utbetaltStoenad: Long,
    val nyBruttoStoenad: Long,
    val differanse: Long,
    val grense: EtteroppgjoerGrenseDto,
    val resultatType: EtteroppgjoerResultatType,
    val harIngenInntekt: Boolean,
    val tidspunkt: Tidspunkt,
    val kilde: Grunnlagsopplysning.Kilde,
    val avkortingForbehandlingId: UUID,
    val avkortingSisteIverksatteId: UUID?,
    val vedtakReferanse: List<Long>?,
)

data class EtteroppgjoerGrenseDto(
    val tilbakekreving: Double,
    val etterbetaling: Double,
    val rettsgebyr: Int,
    val rettsgebyrGyldigFra: LocalDate,
)

enum class EtteroppgjoerResultatType {
    TILBAKEKREVING,
    ETTERBETALING,
    INGEN_ENDRING_MED_UTBETALING,
    INGEN_ENDRING_UTEN_UTBETALING,
}
